elasticdl/python/allreduce/tensorflow_optimizer.py

```diff
@@ -0,0 +1,671 @@
+        backward_passes_per_step=1,
+        average_aggregated_gradients=False,
+        num_groups=0,
+        global_batch_count_per_step=None,
+    ):
+        if name is None:
+            name = "Distributed{}".format(type(optimizer).__name__)
+        super(_DistributedOptimizer, self).__init__(
+            name=name, use_locking=use_locking
+        )
+
+        self._optimizer = optimizer
+        self._allreduce_grads = hvd._make_allreduce_grads_fn(
+            name,
+            device_dense,
+            device_sparse,
+            compression,
+            sparse_as_dense,
+            op,
+            gradient_predivide_factor,
+            num_groups,
+        )
+
+        # *ElasticDL Update*: Always create LocalGradientAggregationHelper
+        # for Elastic training with fixed global batch size.
+        self._agg_helper = LocalGradientAggregationHelper(
+            backward_passes_per_step=backward_passes_per_step,
+            allreduce_func=self._allreduce_grads,
+            sparse_as_dense=sparse_as_dense,
+            average_aggregated_gradients=average_aggregated_gradients,
+            rank=hvd.rank(),
+            optimizer_type=LocalGradientAggregationHelper._OPTIMIZER_TYPE_LEGACY,  # noqa: ignore=E501
+            global_batch_count_per_step=global_batch_count_per_step,
+            op=op,
+        )
+
+    def compute_gradients(self, *args, **kwargs):
+        """Compute gradients of all trainable variables.
+
+        See Optimizer.compute_gradients() for more info.
+
+        In DistributedOptimizer, compute_gradients() is overriden to also
+        allreduce the gradients before returning them.
+        """
+        gradients = self._optimizer.compute_gradients(*args, **kwargs)
+        grads, vars = zip(*gradients)
+        if self._agg_helper:
+            avg_grads = self._agg_helper.compute_gradients(grads)
+        else:
+            avg_grads = self._allreduce_grads(grads)
+        return list(zip(avg_grads, vars))
+
+    def apply_gradients(self, *args, **kwargs):
+        """Calls this same method on the underlying optimizer."""
+        if self._agg_helper:
+            return self._agg_helper.apply_gradients(
+                lambda: self._optimizer.apply_gradients(*args, **kwargs),
+                self._optimizer,
+                *args,
+                **kwargs,
+            )
+
+        return self._optimizer.apply_gradients(*args, **kwargs)
+
+    def get_slot(self, *args, **kwargs):
+        """Calls this same method on the underlying optimizer."""
+        return self._optimizer.get_slot(*args, **kwargs)
+
+    def get_slot_names(self, *args, **kwargs):
+        """Calls this same method on the underlying optimizer."""
+        return self._optimizer.get_slot_names(*args, **kwargs)
+
+    def variables(self, *args, **kwargs):
+        """Calls this same method on the underlying optimizer."""
+        return self._optimizer.variables(*args, **kwargs)
+
+
+def DistributedOptimizer(
```

Comment:
Rename `ElasticDistributedOptimizer`?

Reply:
After discussion, we keep the name `DistributedOptimizer` just as the one in horovod package.